### PR TITLE
Toxify Windows

### PR DESCRIPTION
**Pull request by shayne-fletcher**
Let's see how this goes on Windows now.

### CODE REVIEW

**Comment by shayne-fletcher** [comment]
Cc @garyverhaegen-da

File: triggers/service/BUILD.bazel [hunk 1]
```diff
@@ -73,10 +73,7 @@ da_scala_binary(
 da_scala_test(
     name = "tests",
     srcs = glob(["src/test/scala/**/*.scala"]),
-    data = [
-        ":test-model.dar",
-        "@toxiproxy_nix//:bin/toxiproxy-cmd",
-    ],
+    data = ([":test-model.dar"] + (["@toxiproxy_nix//:bin/toxiproxy-cmd"] if not is_windows else [])),
```
Comment (cocreature) [inline triggers/service/BUILD.bazel:76]: This doesn’t work. You have to set it up as a `dev_env_tool` and then change how you locate runfiles to get it from there.

Comment (shayne-fletcher) [reply triggers/service/BUILD.bazel:76]: Oh. Ok.

File: triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala [hunk 2]
```diff
@@ -60,10 +60,15 @@ object TriggerServiceFixture {
       mat: Materializer,
       aesf: ExecutionSequencerFactory,
       ec: ExecutionContext): Future[A] = {
-    // Launch a toxiproxy instance . Wait on it to be ready to accept
+    // Launch a toxiproxy instance. Wait on it to be ready to accept
     // connections.
     val host = InetAddress.getLoopbackAddress()
-    val toxiProxyExe = BazelRunfiles.rlocation("external/toxiproxy_nix/bin/toxiproxy-cmd")
+    val toxiProxyExe =
+      if (sys.props("os.name").toLowerCase.contains("windows")) {
```
Comment (cocreature) [inline triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala:67]: Please don’t do this. Define it as a `dev_env_tool` like I suggested and locate that.

Comment (shayne-fletcher) [reply triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala:67]: Gimme a break Moritz! I'm working on it!!

Comment (cocreature) [reply triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala:67]: Sorry I got a bit irritated that you just pushed this commit after my original suggestion to use `dev_env_tool` :slightly_smiling_face:

**Comment by cocreature** [comment]
You can fix the error on CI with `    fail_not_supported = False,` in the definition of `toxyproxy_nix`

File: triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala [hunk 3]
```diff
@@ -60,11 +60,12 @@ object TriggerServiceFixture {
       mat: Materializer,
       aesf: ExecutionSequencerFactory,
       ec: ExecutionContext): Future[A] = {
-    // Launch a toxiproxy instance . Wait on it to be ready to accept
+    // Launch a toxiproxy instance. Wait on it to be ready to accept
     // connections.
     val host = InetAddress.getLoopbackAddress()
-    val toxiProxyExe = BazelRunfiles.rlocation("external/toxiproxy_nix/bin/toxiproxy-cmd")
-    val toxiProxyPort = findFreePort()
+    val toxiProxyExe = BazelRunfiles.rlocation("external/toxiproxy_dev_env/toxiproxy")
```
Comment (cocreature) [inline triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala:66]: if it still fails, maybe try adding a `.exe` to the end on Windows or maybe use backslashes.

**Comment by shayne-fletcher** [comment]
tests results just in... chaos monkey can has windows.

Screen Shot 2020-05-27 at 8 53 56 AM

**Comment by shayne-fletcher** [comment]
@cocreature @aherrmann @rohanjr @garyverhaegen-da This appears to work now!

File: triggers/service/BUILD.bazel [hunk 4]
```diff
@@ -73,10 +73,8 @@ da_scala_binary(
 da_scala_test(
     name = "tests",
     srcs = glob(["src/test/scala/**/*.scala"]),
-    data = [
-        ":test-model.dar",
-        "@toxiproxy_nix//:bin/toxiproxy-cmd",
-    ],
+    data = [":test-model.dar"] + (["@toxiproxy_dev_env//:bin/toxiproxy-cmd"] if not is_windows else ["@toxiproxy_dev_env//:toxiproxy-server-windows-amd64.exe"]),
+    jvm_flags = ["-Dcom.daml.toxiproxy=$(rootpath @toxiproxy_dev_env//:bin/toxiproxy-cmd)"] if not is_windows else ["-Dcom.daml.toxiproxy=$(rootpath @toxiproxy_dev_env//:toxiproxy-server-windows-amd64.exe)"],
```
Comment (cocreature) [inline triggers/service/BUILD.bazel:77]: I don’t think you actually need the condition here `$(rootpath @toxiproxy_dev_env//:toxiproxy)` should be sufficient. In Bazel that works fine but the runfiles libraries for Scala and others don’t seem to resolve file groups and aliases properly ime which is probably why just locating it in Scala didn’t work.

File: triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala [hunk 5]
```diff
@@ -60,10 +60,11 @@ object TriggerServiceFixture {
       mat: Materializer,
       aesf: ExecutionSequencerFactory,
       ec: ExecutionContext): Future[A] = {
-    // Launch a toxiproxy instance . Wait on it to be ready to accept
+    // Launch a toxiproxy instance. Wait on it to be ready to accept
     // connections.
     val host = InetAddress.getLoopbackAddress()
-    val toxiProxyExe = BazelRunfiles.rlocation("external/toxiproxy_nix/bin/toxiproxy-cmd")
+    val toxiProxyExe = BazelRunfiles.rlocation(System.getProperty("com.daml.toxiproxy"))
+    println(toxiProxyExe.toString)
```
Comment (cocreature) [inline triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala:67]: Seems like we can remove this now?

File: triggers/service/BUILD.bazel (hunk 4 above)
Comment (shayne-fletcher) [reply triggers/service/BUILD.bazel:77]: we tried many variants of the above. in the end, what you see is the only thing we managed to get to work :)

Comment (aherrmann-da) [reply triggers/service/BUILD.bazel:77]: > I don’t think you actually need the condition here `$(rootpath @toxiproxy_dev_env//:toxiproxy)` should be sufficient.

Trouble is, on Linux this is a `filegroup` containing multiple files so Bazel won't accept `rootpath`. You can use `rootpaths` but then you need to pick the right file path out of those after the fact.

Comment (cocreature) [reply triggers/service/BUILD.bazel:77]: ah makes sense. thanks for the explanation!

**Pull request merged**